azurelinuxagent/ga/exthandlers.py

```diff
@@ -362,9 +360,37 @@ def handle_ext_handlers(self, etag=None):
 
         self.ext_handlers.extHandlers.sort(key=operator.methodcaller('sort_key'))
         for ext_handler in self.ext_handlers.extHandlers:
-            # TODO: handle install in sequence, enable in parallel
+            self.wait_on_ext_handler_dependencies(ext_handler)
             self.handle_ext_handler(ext_handler, etag)
-    
+
+    def wait_on_ext_handler_dependencies(self, ext_handler):
+        dependencies = sum([e.dependencies for e in ext_handler.properties.extensions], [])
+        for dependency in dependencies:
+            handler_i = ExtHandlerInstance(dependency.handler, self.protocol)
+            timeout = 90 if dependency.timeout is None else dependency.timeout
```

Comment:
I worry about the impact on TDP with this change, and possible side effects.  Does this code ever get run if a dependency does not exist?

Reply:
What do you mean by "if a dependency does not exist" ?
If a handler's extensions have no dependencies then this `for` loop should not run.
A missing dependency should be rejected at the template level before the agent even sees it. Even if it does make its way to the agent, `resolve_dependencies` should end up generating a `Dependency` where `exts` is empty so the `for` loop below won't run.

Reply:
Where did 90 seconds come from? We give extensions 5 minutes, iirc? Either way, it should come from the spec.

Reply:
90 *minutes* is the existing timeout for an extension operation at the platform level. That includes the 15 minutes for install, and 5 minutes for enable. In the case of an update, that includes up to 45 minutes (disable + update + uninstall + install + enable).

Extensions have 5 minutes for the enable process to complete. If that process writes out the status blob then this will either return (almost) immediately in the case of a success, or it will timeout in the case of a failure and prevent the installation of further extensions which depend on the extension that failed.

If that process forks a child process which runs beyond the 5 minute timeout, then who knows when the final status blob will get written. Most likely not before the first iteration of this loop for an extension that depends on the extension whose enable command forked a child process.

Reply:
The platform timeout is adjusted to match the number of dependency levels. I have changed the code so that it more closely matches this requirement. It can't match exactly because of the time it takes for the platform to send a new GoalState and the agent to start processing extensions from that new GoalState.